File: src/container/EdgeScoreMap.go
package container

import "github.com/HuXin0817/dots-and-boxes/src/model"

type EdgeScoreMap struct {
	time      [model.MaxEdge]int
	score     [model.MaxEdge]int
	bestEdges EdgeList
}

func (r *EdgeScoreMap) Add(edge model.Edge, score int) {
	r.time[edge]++
	r.score[edge] += score
}

func (r *EdgeScoreMap) Plus(m *EdgeScoreMap) {
	for e := range model.MaxEdge {
		r.time[e] += m.time[e]
		r.score[e] += m.score[e]
	}
}

func (r *EdgeScoreMap) Export() []model.Edge {
	var maxs float64
	for e := range model.MaxEdge {
		if r.time[e] > 0 {
			s := float64(r.score[e]) / float64(r.time[e])
			if s > maxs || r.bestEdges.Empty() {
				maxs = s
				r.bestEdges.Reset(e)
			} else if s == maxs {
				r.bestEdges.Append(e)
			}
		}
	}
	return r.bestEdges.Export()
}

File: src/model/EdgeScoreMap.cpp
#include "EdgeScoreMap.h"

void
EdgeScoreMap::Reset() {
  Time = Array<int, Edge::Max>{};
  Score = Array<int, Edge::Max>{};
  BestEdges.Clear();
}

void
EdgeScoreMap::Add(Edge edge, int score) {
  Time.At(edge)++;
  Score.At(edge) += score;
}

EdgeScoreMap
EdgeScoreMap::operator+=(const EdgeScoreMap& other) {
  for (int i = 0; i < Edge::Max; i++) {
    Time.At(i) += other.Time.At(i);
    Score.At(i) += other.Score.At(i);
  }
  return *this;
}

Span<const Edge>
EdgeScoreMap::Export() {
  float maxs = 0.0;
  for (int e = 0; e < Edge::Max; e++) {
    if (Time.At(e) > 0) {
      float s = static_cast<float>(Score.At(e)) / static_cast<float>(Time.At(e));
      if (s > maxs || BestEdges.Empty()) {
        maxs = s;
        BestEdges.Reset(e);
      } else if (s == maxs) {
        BestEdges.Append(e);
      }
    }
  }
  return BestEdges.Export();
}
